prometheus-mktxp-exporter NixOS module init + mktxp bump

tsandrini opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Sets up a NixOS module for the popular https://github.com/akpw/mktxp prometheus mikrotik exporter :pray:  . This extends the already present work since the mktxp package itself is already prepared in nixpkgs -> https://github.com/NixOS/nixpkgs/blob/dbe18835a88a1572f2ece46eac03be2f920590ea/pkgs/by-name/mk/mktxp/package.nix

~~**Important note**: mktxp only recently added support for loading `username` + `password` via a `credentials_file` secrets option (https://github.com/akpw/mktxp/pull/249), which is why I also bumped the version of the mktxp package in a separate commit, because without this the NixOS module is frankly rather useless in real world scenarios.~~

~~**Edit**: In the meantime mktxp `1.2.14` has been merged into master, so the previous paragraph no longer applies and this PR introduces only one commit -> the mktxp module init.~~

Changes made:

- ~~bumped mktxp from 1.2.9 -> 1.2.11 + also appended myself as an additional maintainer~~
- initialized the NixOS module - there are a few quirks regarding the implementation which I've explained in the accompanied doccomments 
- added a simple NixOS test - more extensive testing would require some sort of mikrotik device mocking
- bumped mktxp to `1.2.15` and included various fixes of the mktxp derivation, for explanation see the comments lower in this PR

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

tsandrini commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 422550`
Commit: `2644434203ce3aeea0d8e26058ef059939dee3fd`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>mktxp</li>
  </ul>
</details>

BonusPlay reviewed (approved):
reviewing just the `mxtxp` package bump, as I'm unfamiliar with the exporter package format

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5701

tsandrini committed:
nixos/prometheus/mktxp: init module

nixos/prometheus/mktxp: update structure, switch to pkgs.formats.ini,etc

nixos/prometheus/mktxp: update test with new settings

nixos/prometheus/mktxp: update options to mktxp version 1.2.15

h7x4 commented on nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix:
```diff
@@ -0,0 +1,236 @@
+    };
+
+    configuration = mkOption {
+      type = types.attrsOf types.anything;
```
If you use `pkgs.formats.ini`, you also get a more accurate type for this as a freeform submodule.

See example usage: https://github.com/NixOS/nixpkgs/blob/81706883222ff3d6142d0e912f6bda7044078d90/nixos/modules/services/system/zram-generator.nix

tsandrini replied:
Thanks for the review! :heart:  I appreciate the feedback! I did some prototyping and I can't decide between two variants of replacing the current `lib.generators.toINI` to `pkgs.format.ini`, do you think that I should

a. Just use a simple `type = lib.types.submodule { freeformType = configurationFormat.type; };` and then merge the resulting configuration with the `defaultConfiguration` that I have already prepared, or
b. I could also prepare explicit options for all of the key/values that are already defined in the `defaultConfiguration`, this could be pretty nice for the end user since the description would should up directly in the NixOS option and if they didn't like them, they could always just pass the `configFile` directly, however, that complicates things a bit, since there are practically 3 different INI sections

1. `MKTXP :: mkMktxpSection` => special section with its structure
2. `default :: mkDeviceSection` => device type with special meaning
3. `* :: mkDeviceSection` => generic mikrotik device, same type as the default section! 

It's kinda hard to create one options submodule that would be 1:1 and also have the documentation and types, etc.. the closest thing that I roughly sketched out is (pasting just the relevant parts, not the entire code)

**First off, define the types**
```nix
mkMktxpSection = lib.types.submodule {
  freeformType = configurationFormat.type; 
  options = {
    listen = lib.mkOption {
      type = lib.types.str;
      default = "${cfg.listenAddress}:${toString cfg.port}";
      description = "Listen address of the mktxp exporter, e.g. 0.0.0.0:49090";
    };
    bandwidth = lib.mkOption {
      type = lib.types.bool; default = false;
      description = "Turns bandwidth metrics collection on/off.";
    };
    # etc...
};

mkDeviceSection = lib.types.submodule {
  freeformType = configurationFormat.type;  
  options = {
    enabled = lib.mkOption {
      type = lib.types.bool; default = false;
      description = "Enable metrics collection for this RouterOS device.";
    };
    hostname = lib.mkOption { type = lib.types.str; default = "localhost"; };
    port     = lib.mkOption { type = lib.types.int; default = 8728; };
    # etc..
  };
};
```

**Then** define the options
```nix
configuration = lib.mkOption {
  type = lib.types.submodule {
    options = {
      # [MKTXP]
      MKTXP = lib.mkOption {
        type = mkMktxpSection;
        default = {};
        description = "Global MKTXP settings (rendered as the [MKTXP] section).";
      };

      # [default]
      default = lib.mkOption {
        type = mkDeviceSection;
        default = {};
        description = "Defaults for all devices (rendered as the [default] section).";
      };

      # [<device-name>]*
      devices = lib.mkOption {
        type = lib.types.attrsOf mkDeviceSection;
        default = {};
        description = ''
          Per-device sections keyed by device name.
          Each attribute becomes a [<name>] INI section.
        '';
      };
    };
  };
  default = {};
  description = "Structured mktxp configuration (typed sections + freeform keys).";
};
```

**The issue with this approach** is that it isn't 1:1 nix to config, however, the tradeoff is we get explicit documentation and typing for the user.

I am not sure, what do you think? Again, thanks for the idea and recommendation :heart: I already implemented the other suggestions you've already mentioned so just meditating on this options thing.

tsandrini replied:
@h7x4 Oh right! Probably did something wrong, I'll try it out again, thanks >.<! Regarding the `configuration -> settings`, that is something I generally follow in my modules, but when designing this module I tried copying the structure/naming scheme of other prometheus exporters and it seems like a lot of them use `configuration` and/or `configFile`

- https://github.com/NixOS/nixpkgs/blob/c6340d0caae1d4d75f02ab2f8c81ad66e8989551/nixos/modules/services/monitoring/prometheus/exporters/dnssec.nix#L15
- https://github.com/NixOS/nixpkgs/blob/c6340d0caae1d4d75f02ab2f8c81ad66e8989551/nixos/modules/services/monitoring/prometheus/exporters/idrac.nix#L36
- https://github.com/NixOS/nixpkgs/blob/c6340d0caae1d4d75f02ab2f8c81ad66e8989551/nixos/modules/services/monitoring/prometheus/exporters/junos-czerwonk.nix#L43
- https://github.com/NixOS/nixpkgs/blob/c6340d0caae1d4d75f02ab2f8c81ad66e8989551/nixos/modules/services/monitoring/prometheus/exporters/mail.nix#L181
- https://github.com/NixOS/nixpkgs/blob/c6340d0caae1d4d75f02ab2f8c81ad66e8989551/nixos/modules/services/monitoring/prometheus/exporters/mikrotik.nix#L32
- etc...

But I'm completely fine with naming it `settings` as well if you think that's better! Thanks for the help!

h7x4 commented on nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix:
```diff
@@ -0,0 +1,236 @@
+    # https://github.com/akpw/mktxp/blob/f2299ee5c264bb1a3d1d2bb781a4e63495e9c641/mktxp/cli/options.py#L233
+    path = [ pkgs.which ];
+
+    # NOTE: mktxp unfortunately does not support passing a custom
+    # configuration file or a custom config directory, it always reads what is
+    # present in $HOME/mktxp/mktxp.conf, so we populate a dynamics user's
+    # home directory with a symlink to the pure store configuration file.
```
I think it might be possible to simplify this a lot if you were to bindmount (see [`systemd.exec(5)`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#BindPaths=)) a derivation over the directory where it looks for config.

h7x4 commented on nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix:
```diff
@@ -0,0 +1,236 @@
+    serviceConfig = {
+      DynamicUser = true;
+      RuntimeDirectory = "mktxp";
+      ExecStart = "${getExe cfg.package} export ${concatStringsSep "\\\n " cfg.extraFlags}";
```
```suggestion
      ExecStart = "${getExe cfg.package} export ${escapeShellArgs cfg.extraFlags}";
```

tsandrini committed:
mktxp: add pyyaml to dependencies

tsandrini committed:
mktxp: add runtime dependency `which`

tsandrini committed:
mktxp: add maintainer tsandrini

tsandrini commented:
@h7x4 So it took me some time to iterate over various setups and properly test them out for a few days in a production settings. I've updated the module accordingly with what we discussed and I am fairly content with the result, been running this setup for two different mikrotik devices and it works well, however, there are a few things to note as well, here is everything that I did/changed:

1. Renamed `cfg.configuration` -> `cfg.settings`
2. `concatStringsSep "\\\n "` -> `escapeShellArgs`
3. Resolved the `path = [ pkgs.which ];`
  - since `mktxp` itself isn't really usable without `which` (i.e. not just within the NixOS module, but the standalone binary) I decided to add the `wrapProgram` call directly to the package definition
  - separate commit [`976a497` (#422550)](https://github.com/NixOS/nixpkgs/pull/422550/commits/976a4973770e8617d20c169c0fb1b1f8463c6a33)
4. Added missing dependency `pyyaml` to `mktxp`
  - since `mktxp` has been updated to `1.2.14` (https://github.com/NixOS/nixpkgs/commit/a014608fd6cf0e65ccb6cde6b1314d977ba9e401) there was now a missing `pyyaml` dependency that is used for loading the `credentials_file` so I've added that
  - separate commit [`7ce26d9` (#422550)](https://github.com/NixOS/nixpkgs/pull/422550/commits/7ce26d97ab93f416f766142026242cf42972474a) + also added myself as a mktxp maintainer
5. Switched from `lib.generators.toINI` to `pkgs.formats.ini`
  - the end user UX is fairly nice
  - I mostly copied the documentation strings 1:1 from the example generated config file, but I tweaked certain words where it made sense (for example instead of specifying a string of comma separated values I just changed it to a list)
  - Also added conversion for nix lists `val = [ a b c ]` to mktxp configuration comma separated lists `val="a, b, c"` https://github.com/NixOS/nixpkgs/blob/d05b8b273c64d6f3ab57e3902d7a67b97e9801cb/nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix#L31
  - Like roughly 90% of the options types are trivially converted, but there are a few special edge cases where I did some special treatment, mktxp isn't completely 100% consistent unfortunately
    - `credentials_file`: mktxp uses "" to indicate no file present, which isn't a valid `path` type so I set the type to `types.nullOr types.path` and added a conversion to string for these https://github.com/NixOS/nixpkgs/blob/d05b8b273c64d6f3ab57e3902d7a67b97e9801cb/nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix#L203
    - `custom_labels`: mktxp uses `Null` here to indicate emtpiness, so `types.nullOr (types.listOf types.str);` for ease of conversions https://github.com/NixOS/nixpkgs/blob/d05b8b273c64d6f3ab57e3902d7a67b97e9801cb/nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix#L209
    - `ssl_ca_file`: again, mktxp uses "" to indicate emptiness, so `types.nullOr types.path;` + conversion https://github.com/NixOS/nixpkgs/blob/d05b8b273c64d6f3ab57e3902d7a67b97e9801cb/nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix#L236
6. Updated prometheus mktxp test to use the new `settings` config https://github.com/NixOS/nixpkgs/blob/d05b8b273c64d6f3ab57e3902d7a67b97e9801cb/nixos/tests/prometheus-exporters.nix#L881
7. Regarding bind mounts
  - I tried a bunch of different variants of using `BindPaths` and `BindReadOnlyPaths` but it just wasn't working well, mktxp really wants a writeable target so it can create a `_mktxp.conf` even if it doesn't get used, I ultimately ended up with the same dynamic user setup but simplified it a bit + removed the `configWriteable` option since it mostly doesn't need to exist anymore

------

Now a have one last note, the current `1.2.14` release has a new bug -> https://github.com/akpw/mktxp/issues/266 ... I am not sure if this happens to everyone but this affected all of my setups and from my analysis it seemed like `mktxp/collector/base_collector.py` needed patching which is what this commit fixed https://github.com/akpw/mktxp/commit/7c25391d4aea8857931351c05f09805a170f423a .... So I am not sure whether we should just leave this be or change the version to an unstable one `version = "1.2.14-unstable-2025-10-12";` and use `main`. What do you think? This would be enough

```patch
diff --git a/pkgs/by-name/mk/mktxp/package.nix b/pkgs/by-name/mk/mktxp/package.nix
index f1b5256f5..71ffbdc90 100644
--- a/pkgs/by-name/mk/mktxp/package.nix
+++ b/pkgs/by-name/mk/mktxp/package.nix
@@ -6,7 +6,7 @@
   which,
 }:
 let
-  version = "1.2.14";
+  version = "1.2.14-unstable-2025-10-12";
 in
 python3Packages.buildPythonApplication {
   pname = "mktxp";
@@ -16,8 +16,8 @@ python3Packages.buildPythonApplication {
   src = fetchFromGitHub {
     owner = "akpw";
     repo = "mktxp";
-    tag = "v${version}";
-    hash = "sha256-4+0aw/r71FcVrxASco3AkYzi7zbFeiEkJB7acGdb1FQ=";
+    rev = "7c25391d4aea8857931351c05f09805a170f423a";
+    hash = "sha256-yEUSa76hduQk8DIa9w9znmCzJxba23pdoIveB12ff8E=";
   };

   nativeBuildInputs = with python3Packages; [
```

Thanks and have a nice day! <3

tsandrini commented:
@BonusPlay Hi, thanks for the look. I can bump it here as well, as I have described in my previous comment, the current `1.2.14` is unusable, bumping it to `1.2.15` would solve the issue + missing `pyyaml` + and also missing `which` runtime dependency. :thinking: Should I do that?

 Though I still need someone to review the new changes in the NixOS module

tsandrini committed:
mktxp: 1.2.14 -> 1.2.15

BonusPlay commented on nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix:
```diff
@@ -0,0 +1,586 @@
+          MKTXP.listen = "${cfg.listenAddress}:${toString cfg.port}";
+        };
+        merged = recursiveUpdate mktxpSettings cfg.settings;
+      in
```
The only nitpick I can come up with is that I would pull this part out into separate variable instead of another `let in` nested in `else`.

tsandrini replied:
The idea is that the structure mimics the evaluation tree of the final configuration file and you can easily see how it gets created but I don't mind changing it :thinking:  It could also be rewritten into something like this

```nix
  configFile =
    if cfg.configFile != null then
      cfg.configFile
    else
      settingsFormat.generate "mktxp.conf" (
        recursiveUpdate {
          MKTXP.listen = "${cfg.listenAddress}:${toString cfg.port}";
        } cfg.settings
      );
```

But yeah, I don't know. I think going forwards `MKTXP.listen` will probably be the only parameter that needs to be handled like this, so we can even shorten it I guess.

I honestly don't have strong opinions about this, I am fine with all of these options and will gladly change it TwT

**NOTE**: Just as an additional explanation, prometheus exporter modules have some default structure with some predefined options, two of those being `listenAddress` and `port`, which is why they are not included in `settings`.

BonusPlay reviewed (approved):
I'm by no means an expert on nixos modules, but this code seems decent. I haven't tested it in practice though.

h7x4 commented on nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix:
```diff
@@ -0,0 +1,586 @@
+      default = null;
+      description = ''
+        The path to the mktxp configuration ini file. For more info see
+        https://github.com/akpw/mktxp
```
```suggestion
        <https://github.com/akpw/mktxp>
```

h7x4 commented on nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix:
```diff
@@ -0,0 +1,586 @@
+          https://github.com/akpw/mktxp
+
+          Note that setting this option is mutually exclusive with setting
+          the `configFile` option.
```
```suggestion
          <https://github.com/akpw/mktxp>

          Note that setting this option is mutually exclusive with setting
          the {option}`configFile` option.
```

h7x4 commented on nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix:
```diff
@@ -0,0 +1,586 @@
+      '';
+    };
+
+    settings =
```
It seems completely overkill to declare and slightly painful to maintain all of these options here. I imagine it could easily lead to a lot of renames/removed option imports over time.

Is it the case that the rendered config file needs to have all of these options in order for the program to accept it? If so, could you instead limit the options to those you know people want to change and/or those with weird type conversions, and then just add the rest of the defaults via the `recursiveUpdate` further up?

h7x4 reviewed (commented):
Added a few comments, none blocking, feel free to just ping me again :)

h7x4 commented on nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix:
```diff
@@ -0,0 +1,586 @@
+      DynamicUser = true;
+      RuntimeDirectory = "mktxp";
+      ExecStart = "${getExe cfg.package} export ${escapeShellArgs cfg.extraFlags}";
+    };
```
I just realized I had reviewed this before, and read your comments:

> Regarding bind mounts

> I tried a bunch of different variants of using BindPaths and BindReadOnlyPaths but it just wasn't working well, mktxp really wants a writeable target so it can create a _mktxp.conf even if it doesn't get used, I ultimately ended up with the same dynamic user setup but simplified it a bit + removed the configWriteable option since it mostly doesn't need to exist anymore

Could you test if the above works? It should leave the outer `/run/mktxp/mktxp/` writable, while keeping the `mktxp.conf` readonly (the same behaviour as before, with the symlink to the nix store)

iliakonnov commented on nixos/modules/services/monitoring/prometheus/exporters/mktxp.nix:
```diff
@@ -0,0 +1,586 @@
+    # NOTE: mktxp unfortunately does not support passing a custom
+    # configuration file or a custom config directory, it always reads what is
+    # present in $HOME/mktxp/mktxp.conf, so we populate a dynamics user's
+    # home directory with a symlink to the pure store configuration file.
```
It looks like specifying custom config directory is possible with `--cfg-dir` flag.

From [README](https://github.com/akpw/mktxp/blob/main/README.md?plain=1#L181-L183):
```
# Run with dedicated config directory
docker run -v "$(pwd)/mktxp-config:/etc/mktxp" -p 49090:49090 -it --rm \
  ghcr.io/akpw/mktxp:main mktxp --cfg-dir /etc/mktxp export
```

However, mktxp creates an `_mktxp.conf` in this directory, so I don't think we can place config directory in Nix store directly.
